File: android/jvm-tests/src/test/kotlin/KeywordsTest.kt
import android.content.Context
import hu.breaker.app.core.AppState
import hu.breaker.app.core.BreakerStore
import hu.breaker.app.core.ChallengeEngine.Step
import hu.breaker.app.core.Referee
import hu.breaker.app.core.Focus
import hu.breaker.app.core.FocusSync
import hu.breaker.app.core.KeywordLogic
import hu.breaker.app.core.SyncClient
import hu.breaker.app.core.SyncRevisions
import org.json.JSONArray
import org.json.JSONObject
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Kulcsszó-szabályok a Kotlin tükrön — a desktop/test/keywords.test.ts esetei:
 * a mag (alak, lista, fésülés, illesztés), a jel, a drót és a mentés — és a
 * bíró: a telefon nem érvényesít (a DNS a címet nem látja), de szerkeszt és
 * hordoz: felvenni ingyen, levenni próbatétel.
 */
class KeywordsTest {

    private val now = 1_700_000_000_000L

    @BeforeTest fun reset() {
        BreakerStore.init(Context())
        BreakerStore.mutate { AppState() }
    }

    @Test fun `a kulcsszo kanonikus alakja - kisbetu, NFKC, szelek le`() {
        assertEquals("shorts", KeywordLogic.normalizeKeyword("  Shorts "))
        assertEquals("reels", KeywordLogic.normalizeKeyword("REELS"))
        assertNull(KeywordLogic.normalizeKeyword("ab"), "két betű mindenre illene")
        assertNull(KeywordLogic.normalizeKeyword("két szó"))
        assertNull(KeywordLogic.normalizeKeyword("x".repeat(KeywordLogic.MAX_KEYWORD_LENGTH + 1)))
        assertEquals("x".repeat(KeywordLogic.MAX_KEYWORD_LENGTH), KeywordLogic.normalizeKeyword("x".repeat(KeywordLogic.MAX_KEYWORD_LENGTH)))
        assertNull(KeywordLogic.normalizeKeyword(null))
        assertNull(KeywordLogic.normalizeKeyword(""))
        assertEquals("álom", KeywordLogic.normalizeKeyword("a" + Char(0x0301) + "lom"))
    }

    @Test fun `a lista tisztan es a kulcs rendezett`() {
        assertEquals(listOf("shorts", "reels"), KeywordLogic.cleanKeywords(listOf("Shorts", "shorts", "ab", "reels", null, " REELS ")))
        val many = (0 until KeywordLogic.MAX_KEYWORDS + 5).map { "szo${it.toString().padStart(3, '0')}" }
        assertEquals(KeywordLogic.MAX_KEYWORDS, KeywordLogic.cleanKeywords(many).size)
        assertEquals(KeywordLogic.keywordsKey(listOf("reels", "shorts")), KeywordLogic.keywordsKey(listOf("shorts", "reels")))
        assertTrue(KeywordLogic.sameKeywords(listOf("Shorts", "reels"), listOf("reels", "shorts")))
        assertFalse(KeywordLogic.isKeywordsLoosening(listOf("shorts"), listOf("shorts", "reels")), "bővítés: szigorítás")
        assertTrue(KeywordLogic.isKeywordsLoosening(listOf("shorts", "reels"), listOf("shorts")), "levétel: lazítás")
    }

    @Test fun `fesules a jel szerint es illesztes a cimre`() {
        assertEquals(listOf("reels"), KeywordLogic.mergeKeywords(3, listOf("shorts"), 5, listOf("reels")))
        assertEquals(listOf("shorts", "reels"), KeywordLogic.mergeKeywords(3, listOf("shorts"), 3, listOf("reels")), "azonos jel: unió")
        assertEquals(listOf("shorts"), KeywordLogic.mergeKeywords(0, listOf(), 0, listOf("shorts")), "a jeltelen nem töröl")
        val words = listOf("shorts", "tiktok", "játék")
        assertEquals("shorts", KeywordLogic.keywordHit(words, "https://www.youtube.com/shorts/abc"))
        assertEquals("shorts", KeywordLogic.keywordHit(words, "https://www.youtube.com/watch?v=x&list=SHORTS"))
        assertEquals("tiktok", KeywordLogic.keywordHit(words, "https://www.tiktok.com/@valaki"), "a hosztnév is a cím része")
        assertEquals("játék", KeywordLogic.keywordHit(words, "https://example.com/j%C3%A1t%C3%A9k"), "a százalék-kódolás feloldva")
        assertNull(KeywordLogic.keywordHit(words, "https://example.com/hirek"))
        assertNull(KeywordLogic.keywordHit(words, "https://example.com/%E0%A4%A"), "rossz kódolás: nem hasal el")
        assertNull(KeywordLogic.keywordHit(listOf("ab"), "https://ab.com"))
    }

    @Test fun `a jel - a lista cserele lepteti a blobot, az atvetel nem`() {
        var st = AppState()
        assertEquals(st.copy(focusRevFp = SyncRevisions.focusFingerprint(st)), SyncRevisions.bumpFocus(st, "telefon", now), "üres: nincs léptetés")
        st = SyncRevisions.bumpFocus(st.copy(keywords = listOf("shorts")), "telefon", now)
        assertEquals(1L, st.focusRev)
        assertEquals(1, st.keywordsRev)
        assertEquals(st, SyncRevisions.bumpFocus(st, "telefon", now + 1), "változatlanul nem léptet")
        st = SyncRevisions.bumpFocus(st.copy(focusPacks = listOf(Focus.FocusPack("p1", "Írás", emptyList(), emptyList(), 25))), "telefon", now + 2)
        assertEquals(2L, st.focusRev)
        assertEquals(1, st.keywordsRev, "a csomag szerkesztése nem a kulcsszavak jele")
        val adopted = SyncRevisions.adoptFocus(st.copy(keywords = listOf("reels"), keywordsRev = 9))
        assertEquals(adopted, SyncRevisions.bumpFocus(adopted, "telefon", now + 3), "az átvétel nem szerkesztés")
        val edited = SyncRevisions.bumpFocus(adopted.copy(focusPacks = emptyList()), "telefon", now + 4)
        assertEquals(9, edited.keywordsRev, "az átvett lista jele marad")
    }

    @Test fun `a drot - a lista es a jele oda-vissza, a szemet kiesik, a fesules a blobon`() {
        val base = FocusSync.SyncFocus(packs = emptyList(), run = null, log = emptyList(), rev = 0, updatedAt = 0, updatedBy = "dev")
        val f = SyncClient.focusFromJson(
            JSONObject().put("packs", JSONArray()).put("rev", 4).put("keywordsRev", 99)
                .put("keywords", JSONArray(listOf("Shorts", "ab", "shorts", "reels"))).toString(),
            "dev", null,
        )
        assertEquals(listOf("shorts", "reels"), f.keywords)
        assertNull(f.keywordsRev, "a jel legfeljebb a blob rev-je")
        val text = SyncClient.focusToJson(base.copy(keywords = listOf("shorts"), keywordsRev = 3, rev = 3))
        assertTrue(text.contains("\"keywords\""))
        assertTrue(text.contains("\"keywordsRev\":3"))
        assertFalse(SyncClient.focusToJson(base).contains("keywords"), "üresen nincs mező")

        val local = base.copy(keywords = listOf("shorts"), keywordsRev = 3, rev = 3)
        val removed = base.copy(keywordsRev = 5, rev = 5, updatedBy = "other")
        assertEquals(emptyList(), FocusSync.merge(local, removed).keywords, "a nagyobb jelű levétel átmegy")
        assertEquals(5, FocusSync.merge(local, removed).keywordsRev)
        assertEquals(listOf("shorts", "reels"), FocusSync.merge(local, base.copy(keywords = listOf("reels"), keywordsRev = 3, rev = 3)).keywords)
        assertEquals(listOf("shorts"), FocusSync.merge(local, base.copy(rev = 9, updatedAt = 999, updatedBy = "old")).keywords, "a jeltelen nem viszi el")
        assertFalse(FocusSync.same(local, local.copy(keywords = listOf("shorts", "reels"))), "a lista cseréje különbség")
    }

    @Test fun `a mentes - a lista, a jel es a kulcs tuleli`() {
        val toJson = BreakerStore::class.java.getDeclaredMethod("toJson", AppState::class.java).apply { isAccessible = true }
        val fromJson = BreakerStore::class.java.getDeclaredMethod("fromJson", JSONObject::class.java).apply { isAccessible = true }
        val st = AppState(keywords = listOf("shorts", "reels"), keywordsRev = 4, focusRevKeywords = "reels|shorts")
        val back = fromJson.invoke(BreakerStore, JSONObject(toJson.invoke(BreakerStore, st).toString())) as AppState
        assertEquals(listOf("shorts", "reels"), back.keywords)
        assertEquals(4, back.keywordsRev)
        assertEquals("reels|shorts", back.focusRevKeywords)
        // A kulcsszavak előtt írt fájl: üres lista, jel nélkül.
        val old = fromJson.invoke(BreakerStore, JSONObject("{\"sites\":[]}")) as AppState
        assertEquals(emptyList(), old.keywords)
        assertNull(old.keywordsRev)
    }

    @Test fun `a javaslatok maguk is ervenyes kulcsszavak`() {
        for (sug in KeywordLogic.SUGGESTIONS) assertEquals(sug, KeywordLogic.normalizeKeyword(sug), sug)
        assertEquals(KeywordLogic.SUGGESTIONS.size, KeywordLogic.SUGGESTIONS.toSet().size)
        assertEquals(KeywordLogic.SUGGESTIONS, KeywordLogic.cleanKeywords(KeywordLogic.SUGGESTIONS))
    }

    // ---------------------------------------------------------------- a bíró

    @Test fun `a biro - felvenni es boviteni ingyen, levenni probatetel - a levetel a teljesiteskor lep eletbe`() {
        assertTrue(Referee.setKeywords(listOf("Shorts"), now).applied, "felvétel ingyen")
        assertEquals(listOf("shorts"), BreakerStore.state.value.keywords)
        assertTrue(Referee.setKeywords(listOf("shorts", "reels"), now).applied, "bővítés ingyen")
        assertTrue(Referee.setKeywords(listOf("reels", "shorts"), now).applied, "ugyanaz más sorrendben: nincs mit tenni")
        assertNull(BreakerStore.state.value.session, "egyik sem indított próbatételt")

        val r = Referee.setKeywords(listOf("shorts"), now)
        assertFalse(r.applied, "levétel: próbatétel")
        assertEquals("keywords", r.session?.siteId)
        assertEquals(listOf("shorts"), r.session?.pendingKeywords)
        assertEquals(listOf("shorts", "reels"), BreakerStore.state.value.keywords, "amíg a próbatétel tart, a lista marad")
        assertEquals("BUSY", assertFailsWith<Referee.RefereeException> { Referee.setKeywords(emptyList(), now) }.code)
        // Futó levétel közben a felvétel ingyen — és a függő lista is tud róla,
        // különben a teljesítéskor a régi lista ülne vissza, és a live eltűnne.
        assertTrue(Referee.setKeywords(listOf("shorts", "reels", "live"), now).applied, "közben a live ingyen")
        assertEquals(listOf("shorts", "live"), BreakerStore.state.value.session?.pendingKeywords)
        solveWholeSession(now)
        assertNull(BreakerStore.state.value.session, "a kísérlet végigment")
        assertEquals(listOf("shorts", "live"), BreakerStore.state.value.keywords, "a reels lement, a live megmaradt")
        assertEquals(1, BreakerStore.state.value.unlockLog.size, "a lazítás a naplóban")
        // A mentés a függő listát is hordozza: egy újraindítás nem tenné feloldássá.
        val again = Referee.setKeywords(listOf("live"), now + 1000)
        assertFalse(again.applied)
        val toJson = BreakerStore::class.java.getDeclaredMethod("toJson", AppState::class.java).apply { isAccessible = true }
        val fromJson = BreakerStore::class.java.getDeclaredMethod("fromJson", JSONObject::class.java).apply { isAccessible = true }
        val back = fromJson.invoke(BreakerStore, JSONObject(toJson.invoke(BreakerStore, BreakerStore.state.value).toString())) as AppState
        assertEquals(listOf("live"), back.session?.pendingKeywords)
    }

    @Test fun `a biro - rossz kulcsszo es tul sok kulcsszo hiba, nem csendes csonkolas`() {
        for (bad in listOf(listOf("ab"), listOf("két szó"), listOf("shorts", "SHORTS"))) {
            assertEquals("BAD_KEYWORD", assertFailsWith<Referee.RefereeException> { Referee.setKeywords(bad, now) }.code, "$bad")
        }
        val many = (0 until KeywordLogic.MAX_KEYWORDS + 1).map { "szo" + it.toString().padStart(3, '0') }
        assertEquals("TOO_MANY_KEYWORDS", assertFailsWith<Referee.RefereeException> { Referee.setKeywords(many, now) }.code)
        assertTrue(BreakerStore.state.value.keywords.isEmpty(), "hibánál semmi nem változik")
        assertNull(BreakerStore.state.value.session)
    }

    /** Végigviszi a futó kísérletet — a várakozó lépést a célpontja után veszi át. */
    private fun solveWholeSession(now: Long) {
        var guard = 0
        while (BreakerStore.state.value.session != null && guard++ < 200) {
            val s = BreakerStore.state.value.session!!
            when (val step = s.steps[s.stepIndex]) {
                is Step.Delay -> Referee.claimDelay(s.id, (step.claimableAt ?: 0) + 1)
                else -> Referee.submitAnswer(s.id, solveStep(step, now), now)
            }
        }
    }

    /** A helyes válasz; a MEMORY lépést visszadátumozzuk a várakozása mögé. */
    private fun solveStep(step: Step, now: Long): String = when (step) {
        is Step.Transcribe -> step.text
        is Step.MathChain -> step.problems[step.pos].a.toString()
        is Step.Memory -> {
            BreakerStore.mutate { st ->
                val ses = st.session!!
                val steps = ses.steps.toMutableList()
                steps[ses.stepIndex] = step.copy(armedAt = now - step.showMs - step.waitMs - 1000)
                st.copy(session = ses.copy(steps = steps))
            }
            step.code
        }
        is Step.Reverse -> step.text.reversed()
        is Step.Delay -> error("a várakozást átvenni kell")
        is Step.Partner -> error("a megbízott lépése a jelmondat — ezek a tesztek megbízott nélkül futnak")
    }
}

File: desktop/test/keywords.test.ts
// Kulcsszó-szabályok: a mag (alak, lista, fésülés, illesztés), a bíró (felvenni
// ingyen, levenni próbatétel), a jel (lenyomat), és a drót (fésülés a blobon).

import { test } from 'node:test';
import * as assert from 'node:assert/strict';
import {
  KEYWORD_SUGGESTIONS, MAX_KEYWORDS, MAX_KEYWORD_LENGTH, MIN_KEYWORD_LENGTH, cleanKeywords, isKeywordsLoosening,
  keywordHit, keywordsKey, mergeKeywords, normalizeKeyword, sameKeywords,
} from '../src/shared/keywords';
import { defaultState, newId, type HelperState } from '../src/helper/state';
import * as referee from '../src/helper/referee';
import { adoptFocusRevision, bumpFocusRevision } from '../src/helper/revisions';
import { mergeFocus, normalizeSyncFocus, sameFocus, type SyncFocus } from '../src/shared/sync/focus-merge';
import { reverseString, type Step } from '../src/shared/challenges';

test('a kulcsszó kanonikus alakja: kisbetű, NFKC, szélek le — szóköz és rövid nem szabály', () => {
  assert.equal(normalizeKeyword('  Shorts '), 'shorts');
  assert.equal(normalizeKeyword('REELS'), 'reels');
  assert.equal(normalizeKeyword('ab'), null, 'két betű mindenre illene');
  assert.equal(normalizeKeyword('két szó'), null, 'egy cím sem tartalmaz szóközt');
  assert.equal(normalizeKeyword('x'.repeat(MAX_KEYWORD_LENGTH + 1)), null);
  assert.equal(normalizeKeyword('x'.repeat(MAX_KEYWORD_LENGTH)), 'x'.repeat(MAX_KEYWORD_LENGTH));
  assert.equal(normalizeKeyword('x'.repeat(MIN_KEYWORD_LENGTH)), 'xxx');
  assert.equal(normalizeKeyword(42), null);
  assert.equal(normalizeKeyword(''), null);
  // Bontott ékezet ugyanaz, mint az összetett.
  assert.equal(normalizeKeyword('a' + String.fromCharCode(0x0301) + 'lom'), 'álom');
});

test('a lista tisztán: csak az érvényes, egyszer, a plafonig; a kulcs rendezett', () => {
  assert.deepEqual(cleanKeywords(['Shorts', 'shorts', 'ab', 'reels', 42, null, ' REELS ']), ['shorts', 'reels']);
  assert.deepEqual(cleanKeywords('nem tömb'), []);
  const many = Array.from({ length: MAX_KEYWORDS + 5 }, (_, i) => `szo${i.toString().padStart(3, '0')}`);
  assert.equal(cleanKeywords(many).length, MAX_KEYWORDS);
  assert.equal(keywordsKey(['reels', 'shorts']), keywordsKey(['shorts', 'reels']), 'a sorrend nem jelentés');
  assert.ok(sameKeywords(['Shorts', 'reels'], ['reels', 'shorts']));
  assert.ok(!sameKeywords(['shorts'], ['shorts', 'reels']));
});

test('lazítás-e: ami a mostaniból hiányzik az újból, az levétel; a bővítés nem', () => {
  assert.equal(isKeywordsLoosening([], ['shorts']), false, 'felvétel: szigorítás');
  assert.equal(isKeywordsLoosening(['shorts'], ['shorts', 'reels']), false, 'bővítés: szigorítás');
  assert.equal(isKeywordsLoosening(['shorts', 'reels'], ['shorts']), true, 'levétel: lazítás');
  assert.equal(isKeywordsLoosening(['shorts'], []), true);
  assert.equal(isKeywordsLoosening(['shorts'], ['SHORTS']), false, 'ugyanaz más alakban');
});

test('fésülés a jel szerint: nagyobb jel nyer, azonos jelnél a bővebb — a jeltelen sosem töröl', () => {
  assert.deepEqual(mergeKeywords(3, ['shorts'], 5, ['reels']), ['reels'], 'a nagyobb jelű levétel átmegy');
  assert.deepEqual(mergeKeywords(5, ['shorts'], 3, ['reels']), ['shorts']);
  assert.deepEqual(mergeKeywords(3, ['shorts'], 3, ['reels']), ['shorts', 'reels'], 'azonos jel: unió');
  assert.deepEqual(mergeKeywords(0, [], 0, ['shorts']), ['shorts'], 'a jeltelen blob nem törli a listát');
  assert.deepEqual(mergeKeywords(2, ['shorts'], 0, []), ['shorts']);
});

test('illesztés: a cím bárhol tartalmazza — hosztban is —, kódolás feloldva, kisbetűvel', () => {
  const words = ['shorts', 'tiktok', 'játék'];
  assert.equal(keywordHit(words, 'https://www.youtube.com/shorts/abc'), 'shorts');
  assert.equal(keywordHit(words, 'https://www.youtube.com/watch?v=x&list=SHORTS'), 'shorts', 'kisbetű-független');
  assert.equal(keywordHit(words, 'https://www.tiktok.com/@valaki'), 'tiktok', 'a hosztnév is a cím része');
  assert.equal(keywordHit(words, 'https://example.com/j%C3%A1t%C3%A9k'), 'játék', 'a százalék-kódolás feloldva');
  assert.equal(keywordHit(words, 'https://example.com/hirek'), null);
  assert.equal(keywordHit(words, 'https://example.com/%E0%A4%A'), null, 'rossz kódolás: marad, nem hasal el');
  assert.equal(keywordHit([], 'https://www.youtube.com/shorts/abc'), null);
  assert.equal(keywordHit(['ab'], 'https://ab.com'), null, 'érvénytelen kulcsszó nem illik');
});

// ------------------------------------------------------------------ a bíró

function stateWithSite(): { state: HelperState; siteId: string } {
  const state = defaultState();
  const siteId = newId('site');
  state.sites.push({
    id: siteId, domain: 'youtube.com', hostnames: ['youtube.com'], addedAt: 1, pauseUntil: null, pendingDeleteAt: null,
  });
  return { state, siteId };
}

function solveStep(step: Step, now: number): string {
  switch (step.type) {
    case 'TRANSCRIBE': return step.text;
    case 'MATH_CHAIN': return String(step.problems[step.pos].a);
    case 'MEMORY': step.armedAt = now - step.showMs - step.waitMs - 1000; return step.code;
    case 'REVERSE': return reverseString(step.text);
    case 'DELAY': throw new Error('a várakozást átvenni kell');
    case 'PARTNER': throw new Error('a megbízott lépése a jelmondat');
  }
}

function solveAll(state: HelperState, now: number): void {
  let guard = 0;
  while (state.session && guard++ < 60) {
    const s = state.session;
    const step = s.steps[s.stepIndex];
    if (step.type === 'DELAY') {
      referee.claimDelay(state, s.id, (step.claimableAt ?? now) + 1);
    } else {
      referee.submitAnswer(state, s.id, solveStep(step, now), now);
    }
  }
  assert.equal(state.session, null, 'a kísérlet végigment');
}

test('a bíró: felvenni és bővíteni ingyen, levenni próbatétel — a levétel a teljesítéskor lép életbe', () => {
  const { state } = stateWithSite();
  const now = 1_700_000_000_000;
  assert.equal(referee.setKeywords(state, ['Shorts'], now).applied, true, 'felvétel ingyen');
  assert.deepEqual(state.keywords, ['shorts']);
  assert.equal(referee.setKeywords(state, ['shorts', 'reels'], now).applied, true, 'bővítés ingyen');
  assert.deepEqual(state.keywords, ['shorts', 'reels']);
  assert.equal(referee.setKeywords(state, ['reels', 'shorts'], now).applied, true, 'ugyanaz más sorrendben: nincs mit tenni');

  const r = referee.setKeywords(state, ['shorts'], now);
  assert.equal(r.applied, false, 'levétel: próbatétel');
  assert.equal(state.session?.siteId, 'keywords');
  assert.deepEqual(state.session?.pendingKeywords, ['shorts']);
  assert.equal(r.session?.keywords, true, 'a felület ebből tudja, mi a tét');
  assert.deepEqual(state.keywords, ['shorts', 'reels'], 'amíg a próbatétel tart, a lista marad');
  assert.throws(() => referee.setKeywords(state, [], now), (e: referee.RefereeError) => e.code === 'BUSY');
  solveAll(state, now);
  assert.deepEqual(state.keywords, ['shorts'], 'a teljesítéskor lép életbe');
  assert.equal(state.unlockLog.length, 1, 'a lazítás a naplóban');

  // Az utolsó levétele: üres lista — és üresen nincs mező.
  const last = referee.setKeywords(state, [], now + 1000);
  assert.equal(last.applied, false);
  solveAll(state, now + 1000);
  assert.equal(state.keywords, undefined);
});

test('a bíró: futó levétel közben a felvétel ingyen — és a levétel végén sem vész el', () => {
  // A szigorítás sosem vár a lazításra: közben is felvehető egy szó. De a
  // teljesítéskor a függő lista ül vissza — ha az nem tudna a közben jött
  // szóról, a szigorítás némán eltűnne, pont a levétel árán.
  const { state } = stateWithSite();
  const now = 1_700_000_000_000;
  referee.setKeywords(state, ['shorts', 'reels'], now);
  assert.equal(referee.setKeywords(state, ['reels'], now).applied, false, 'a shorts levétele próbatétel');
  assert.equal(referee.setKeywords(state, ['shorts', 'reels', 'live'], now).applied, true, 'közben a live ingyen');
  assert.deepEqual(state.keywords, ['shorts', 'reels', 'live']);
  assert.deepEqual(state.session?.pendingKeywords, ['reels', 'live'], 'a függő lista is tud a live-ról');
  solveAll(state, now);
  assert.deepEqual(state.keywords, ['reels', 'live'], 'a shorts lement, a live megmaradt');
});

test('a bíró: rossz kulcsszó és túl sok kulcsszó hiba, nem csendes csonkolás', () => {
  const { state } = stateWithSite();
  const now = 1_700_000_000_000;
  assert.throws(() => referee.setKeywords(state, ['ab'], now), (e: referee.RefereeError) => e.code === 'BAD_KEYWORD');
  assert.throws(() => referee.setKeywords(state, ['két szó'], now), (e: referee.RefereeError) => e.code === 'BAD_KEYWORD');
  assert.throws(() => referee.setKeywords(state, ['shorts', 'SHORTS'], now), (e: referee.RefereeError) => e.code === 'BAD_KEYWORD');
  const many = Array.from({ length: MAX_KEYWORDS + 1 }, (_, i) => `szo${i.toString().padStart(3, '0')}`);
  assert.throws(() => referee.setKeywords(state, many, now), (e: referee.RefereeError) => e.code === 'TOO_MANY_KEYWORDS');
  assert.equal(state.keywords, undefined, 'hibánál semmi nem változik');
});

test('a jel: a felvétel és a levétel lépteti a blobot, az átvétel nem', () => {
  const state = defaultState();
  const now = 1_700_000_000_000;
  assert.equal(bumpFocusRevision(state, 'dev', now), false, 'üres állapot: nincs mit léptetni');
  referee.setKeywords(state, ['shorts'], now);
  assert.equal(bumpFocusRevision(state, 'dev', now), true);
  assert.equal(state.keywordsRev, state.focusRev);
  const rev = state.focusRev!;
  assert.equal(bumpFocusRevision(state, 'dev', now + 1), false, 'változatlan: nem léptet');
  state.focusPacks = [{ id: 'p1', name: 'Írás', allowSites: [], allowApps: [], defaultMinutes: 25 }];
  assert.equal(bumpFocusRevision(state, 'dev', now + 2), true);
  assert.equal(state.keywordsRev, rev, 'a csomag szerkesztése nem a kulcsszavak jele');
  referee.setKeywords(state, ['shorts', 'reels'], now + 3);
  assert.equal(bumpFocusRevision(state, 'dev', now + 3), true);
  assert.equal(state.keywordsRev, rev + 2);
  // Egy másik eszközről átvett lista: a lenyomat és a kulcs újraszámolva — nincs léptetés,
  // és a következő saját szerkesztés sem bélyegzi át a jelét.
  state.keywords = ['reels'];
  state.keywordsRev = 9;
  adoptFocusRevision(state);
  assert.equal(bumpFocusRevision(state, 'dev', now + 4), false, 'az átvétel nem szerkesztés');
  state.focusPacks = [];
  assert.equal(bumpFocusRevision(state, 'dev', now + 5), true);
  assert.equal(state.keywordsRev, 9, 'az átvett lista jele marad');
});

// ------------------------------------------------------------------ a drót

const focus = (over: Partial<SyncFocus>): SyncFocus =>
  ({ packs: [], run: null, log: [], rev: 0, updatedAt: 0, updatedBy: 'x', ...over });

test('a blob hordozza a kulcsszavakat és a jelüket; a szemét és a túl nagy jel kiesik', () => {
  const n = normalizeSyncFocus({ keywords: ['Shorts', 'ab', 'shorts', 42], keywordsRev: 2, rev: 3 }, 'dev');
  assert.deepEqual(n.keywords, ['shorts']);
  assert.equal(n.keywordsRev, 2);
  const big = normalizeSyncFocus({ keywords: ['shorts'], keywordsRev: 99, rev: 3 }, 'dev');
  assert.equal(big.keywordsRev, undefined, 'a jel legfeljebb a blob rev-je');
  const none = normalizeSyncFocus({ keywords: [], rev: 3 }, 'dev');
  assert.equal(none.keywords, undefined, 'üresen nincs mező');
});

test('a blob fésülése: a jel dönt, azonos jelnél az unió; a régi kliens nem töröl', () => {
  const local = focus({ keywords: ['shorts'], keywordsRev: 3, rev: 3 });
  const removed = focus({ keywordsRev: 5, rev: 5, updatedBy: 'other' });
  const merged = mergeFocus(local, removed);
  assert.equal(merged.keywords, undefined, 'a nagyobb jelű levétel átmegy');
  assert.equal(merged.keywordsRev, 5);
  const same = mergeFocus(local, focus({ keywords: ['reels'], keywordsRev: 3, rev: 3 }));
  assert.deepEqual(same.keywords, ['shorts', 'reels'], 'azonos jel: unió');
  const old = focus({ rev: 9, updatedAt: 999, updatedBy: 'old' });
  assert.deepEqual(mergeFocus(local, old).keywords, ['shorts'], 'a jeltelen blob nem viszi el');
  assert.ok(!sameFocus(local, focus({ ...local, keywords: ['shorts', 'reels'] })), 'a lista cseréje különbség: fel kell tölteni');
});

test('a javaslatok maguk is érvényes kulcsszavak — kanonikus alakban, egyszer, a plafon alatt', () => {
  for (const s of KEYWORD_SUGGESTIONS) assert.equal(normalizeKeyword(s), s, s);
  assert.equal(new Set(KEYWORD_SUGGESTIONS).size, KEYWORD_SUGGESTIONS.length);
  assert.ok(KEYWORD_SUGGESTIONS.length <= MAX_KEYWORDS);
  assert.deepEqual(cleanKeywords(KEYWORD_SUGGESTIONS), KEYWORD_SUGGESTIONS);
});
